### PR TITLE
EString

### PR DESCRIPTION
**Pull request by valeriisydorov**

### CODE REVIEW

File: EString/EString.cpp [hunk 1]
```diff
@@ -0,0 +1,1027 @@
...
+EString::EString(const value_type* str) : EString(str, std::strlen(str)) {}
+
+EString::EString(const value_type* str, size_type len) : EString(str, str + len) {}
+
+EString::EString(size_type len, value_type ch) : value(nullptr), value_length(len), capacity_length(len) {
+    value = allocator.allocate(len + 1);
+    for (size_type i = 0; i < len; ++i) {
+        value[i] = ch;
+    }
+    value[len] = '\0';
```
Comment (valeriisydorov) [inline EString/EString.cpp:15]:
```suggestion
    value[len] = 0;
```

File: EString/EString.h [hunk 2]
```diff
@@ -0,0 +1,340 @@
+#ifndef ESTRING_H
+#define ESTRING_H
```
Comment (KyrylSydorov) [inline EString/EString.h]:
```suggestion
#pragma once
```

File: EString/EString.cpp [hunk 3]
```diff
@@ -0,0 +1,1027 @@
...
+        *this = EString(rhs.data(), rhs.size());
+    }
+    return *this;
+}
+
+EString& EString::operator=(EString&& rhs) noexcept {
+    if (this != &rhs) {
+        value = nullptr;
+        value_length = 0;
+        capacity_length = 0;
```
Comment (KyrylSydorov) [inline EString/EString.cpp]: not empty string

File: EString/EString.cpp [hunk 4]
```diff
@@ -0,0 +1,1027 @@
...
+
+EString::EString(const std::string& str) : EString(str.begin(), str.end()) {}
+
+EString::EString(EString&& other) noexcept : EString() {
+    swap(other);
+}
+
+EString& EString::operator=(const EString& rhs) {
+    if (this != &rhs) {
+        *this = EString(rhs.data(), rhs.size());
```
Comment (KyrylSydorov) [inline EString/EString.cpp:30]: Unoptimized

File: EString/EString.cpp [hunk 5]
```diff
@@ -0,0 +1,1027 @@
...
+EString& EString::assign(const value_type* str, size_type len) {
+    return assign(str, str + len);
+}
+
+EString& EString::assign(const value_type* str) {
+    return assign(str, std::strlen(str));
+}
+
+EString& EString::assign(size_type len, value_type ch) {
+    char str[len + 1];
```
Comment (KyrylSydorov) [inline EString/EString.cpp]:
```suggestion
    value_type str[len + 1];
```

File: EString/EString.cpp [hunk 6]
```diff
@@ -0,0 +1,1027 @@
...
+EString& EString::assign(std::initializer_list<value_type> lst) noexcept {
+    return assign(lst.begin(), lst.end());
+}
+
+EString& EString::assign(const std::string& str) {
+    return operator=(str);
+}
+
+EString& EString::assign(const std::string& str, size_type pos, size_type count) {
+    size_type len = str.size();
```
Comment (KyrylSydorov) [inline EString/EString.cpp]:
```suggestion
    const size_type len = str.size();
```

File: EString/EString.cpp [hunk 7]
```diff
@@ -0,0 +1,1027 @@
...
+        value_type* new_value = allocator.allocate(value_length + 1);
+        std::move(value, value + value_length + 1, new_value);
+        allocator.deallocate(value, capacity_length + 1);
+        value = new_value;
+        capacity_length = value_length;
+    }
+}
+
+EString& EString::insert(size_type index, size_type count, value_type ch) {
+    char str[count + 1];
```
Comment (KyrylSydorov) [inline EString/EString.cpp]:
```suggestion
    value_type str[count + 1];
```

File: EString/EString.cpp [hunk 8]
```diff
@@ -0,0 +1,1027 @@
...
+void EString::push_back(value_type ch) {
+    if (value_length + 1 > capacity_length) {
+        reserve(value_length + 1);
+    }
+    value[value_length] = ch;
+    value[++value_length] = '\0';
+}
+
+void EString::pop_back() {
+    value[--value_length] = '\0';
```
Comment (KyrylSydorov) [inline EString/EString.cpp]: exception

File: EString/EString.cpp [hunk 9]
```diff
@@ -0,0 +1,1027 @@
...
+    }
+    size_type avail_count = value_length - pos;
+    size_type real_count = count < avail_count ? count : avail_count;
+    for (size_type i = 0; i < real_count; ++i) {
+        dest[i] = value[pos + i];
+    }
+    return real_count;
+}
+
+void EString::resize(size_type count, char ch) noexcept {
```
Comment (KyrylSydorov) [inline EString/EString.cpp]:
```suggestion
void EString::resize(size_type count, value_type ch) noexcept {
```

File: EString/EString.cpp [hunk 10]
```diff
@@ -0,0 +1,1027 @@
...
+    if (count <= value_length) {
+        value[count] = '\0';
+        value_length = count;
+    } else {
+        reserve(count);
+        for (size_type i = 0; i < count - value_length; ++i) {
+            value[value_length + i] = ch;
+        }
+        value[count] = '\0';
+        value_length = count;
```
Comment (KyrylSydorov) [inline EString/EString.cpp]: move out from this if/else

File: EString/EString.cpp [hunk 11]
```diff
@@ -0,0 +1,1027 @@
...
+}
+
+std::ostream& operator<<(std::ostream& os, const EString& str) {
+    os << str.data();
+    return os;
+}
+
+std::istream& operator>>(std::istream& is, EString& str) {
+    str.clear();
+    char ch;
```
Comment (KyrylSydorov) [inline EString/EString.cpp]:
```suggestion
    EString::value_type ch;
```

File: EString/EString.cpp [hunk 12]
```diff
@@ -0,0 +1,1027 @@
...
+    return EString(lhs) + EString(rhs);
+}
+
+EString operator+(EString&& lhs, EString&& rhs) {
+    EString result(std::move(lhs));
+    return result.append(std::move(rhs));
+}
+
+EString operator+(EString&& lhs, const EString& rhs) {
+    return EString(lhs) + EString(rhs);
```
Comment (KyrylSydorov) [inline EString/EString.cpp]:
```suggestion
    return EString(std::move(lhs)) + EString(rhs);
```

**Review by KyrylSydorov — CHANGES_REQUESTED** [review]

File: EString/main.cpp [hunk 13]
```diff
@@ -1,6 +1,532 @@
...
+    EString es_2 = "David Gries";
+
+    // EString(const value_type*, size_type)
+    EString es_3("Thomas Cormen ", 13);
+
+    // EString(std::initializer_list<value_type>)
+    EString es_4({'D', 'o', 'n', 'a', 'l', 'd', ' ', 'K', 'n', 'u', 't', 'h'});
+
+    // EString(It, It), begin(), end()
+    char arr_5[] = {'D', 'o', 'n', 'a', 'l', 'd', ' ', 'K', 'n', 'u', 't', 'h'};
```
Comment (KyrylSydorov) [inline EString/main.cpp]:
```suggestion
    char arr_5[] = {'D', 'o', 'n', 'a', 'l', 'd', ' ', 'K', 'n', 'u', 't', 'h'};
    char arr_5[] = "...";
```

**Pull request closed without merging**